src/typing.py

```diff
@@ -991,14 +968,26 @@ def __new__(cls, name, bases, namespace,
         initial_bases = bases
         if extra is not None and type(extra) is abc.ABCMeta and extra not in bases:
             bases = (extra,) + bases
-        bases = tuple(_gorg(b) if isinstance(b, GenericMeta) else b for b in bases)
+        new_bases = []
+        for base in bases:
+            if isinstance(base, GenericMeta):
+                bextra = getattr(base, '__extra__', None)
+                if (extra and bextra and not origin and bextra not in bases and
+                        type(bextra) is abc.ABCMeta):
```

Comment:
I am actually a bit worried maybe this is too aggressive, since sometimes we even lose ``Generic``, for example:
```python
>>> pprint(MutableMapping.__mro__)
(typing.MutableMapping,
 <class 'collections.abc.MutableMapping'>,
 <class 'collections.abc.Mapping'>,
 <class 'collections.abc.Collection'>,
 <class 'collections.abc.Sized'>,
 <class 'collections.abc.Iterable'>,
 <class 'collections.abc.Container'>,
 <class 'object'>)
```
Although all tests pass, and I tried to play with this and it works as expected (it is still an instance of ``GenericMeta``). Probably, I will add ``Generic`` at the end of bases in such cases (just for consistency).